app/src/main/java/com/google/samples/apps/sunflower/PlantDetailFragment.kt

```diff
@@ -42,22 +46,42 @@ class PlantDetailFragment : Fragment() {
         val plantId = requireNotNull(arguments).getString(ARG_ITEM_ID)
 
         val factory = InjectorUtils.providePlantDetailViewModelFactory(requireActivity(), plantId)
-        val plantDetailViewModel = ViewModelProviders.of(this, factory)
-                .get(PlantDetailViewModel::class.java)
+        plantDetailViewModel = ViewModelProviders.of(this, factory)
+            .get(PlantDetailViewModel::class.java)
 
         val binding = DataBindingUtil.inflate<FragmentPlantDetailBinding>(
                 inflater, R.layout.fragment_plant_detail, container, false).apply {
             viewModel = plantDetailViewModel
             setLifecycleOwner(this@PlantDetailFragment)
             fab.setOnClickListener { view ->
-                plantDetailViewModel.addPlantToGarden()
-                Snackbar.make(view, R.string.added_plant_to_garden, Snackbar.LENGTH_LONG).show()
+                val isPlanted = plantDetailViewModel.isPlanted.value ?: false
+                if (isPlanted) {
+                    plantDetailViewModel.removePlantFromGarden()
+                    Snackbar.make(view, R.string.removed_plant_from_garden, Snackbar.LENGTH_LONG).show()
+                } else {
+                    plantDetailViewModel.addPlantToGarden()
+                    Snackbar.make(view, R.string.added_plant_to_garden, Snackbar.LENGTH_LONG).show()
+                }
+
             }
         }
 
+        subscribeUi(binding.root)
+
         return binding.root
     }
 
+    private fun subscribeUi(view: View) {
```

Comment:
Can you do this logic using data binding instead?

Reply:
OK, I refactor the part by data binding.